### ISSUE TITLE
Bug: Types are broken in 9.20.1

### ISSUE THREAD

**Issue by Jason3S** [eslint/eslint#19441]
> The types are still broken. `context.report(report);` doesn't accept a `ReportDescriptor`.
> 
> ```ts
> import type { Rule } from 'eslint';
> 
> export function getReport(context: Rule.RuleContext) {
>   const code = context.sourceCode;
>   const report: Rule.ReportDescriptor = {
>     messageId: '42',
>     loc: code.getLocFromIndex(0),
>   };
>   context.report(report); // <--- Error
> }
> ```
>  

 _Originally posted by @Jason3S in [#19414](https://github.com/eslint/eslint/issues/19414#issuecomment-2668987191)_

See: https://stackblitz.com/edit/stackblitz-starters-m8ciayku?file=index.ts

**Comment by Jason3S** [comment]
@fasttime,

Thank you for the quick fix with `code.getLocFromIndex`. I think this issue is also related to the same thing.

**Comment by fasttime** [comment]
Thanks for the issue. This looks more related to the type definitions in `@eslint/core` than in ESLint itself. I can see several problems in the expected typing of the argument of `context.report()`:

**`fix`**

The type of `report.fix` is currently defined [here](https://github.com/eslint/rewrite/blob/e9a987e8d92a6383e9782332e515aa2a719a18af/packages/core/src/types.ts#L479) and [here](https://github.com/eslint/rewrite/blob/e9a987e8d92a6383e9782332e515aa2a719a18af/packages/core/src/types.ts#L513) as:

```ts
    fix?(fixer: RuleTextEditor): RuleTextEdit | Iterable<RuleTextEdit> | null;
```

Unfortunately, this definition does not account for the (legitimate) case where `fix` is itself `null`, as opposed to a function or `undefined`. I'm also not sure if a `fix` function should be allowed to return `undefined` additionally to `null`.

**`suggest`**

The type of `report.suggest` [here](https://github.com/eslint/rewrite/blob/e9a987e8d92a6383e9782332e515aa2a719a18af/packages/core/src/types.ts#L485) should probably allow for `null` values in order to match the definition of `ReportDescriptorOptions`:

https://github.com/eslint/eslint/blob/274f054f19f5f490d9496c6eee4bcd8620d2f4be/lib/types/index.d.ts#L828-L830

**`loc`**

The type of `report.loc` [here](https://github.com/eslint/rewrite/blob/e9a987e8d92a6383e9782332e515aa2a719a18af/packages/core/src/types.ts#L491) is currently:

```ts
    type ViolationLocation<Node> = { loc: SourceLocation } | { node: Node };
```

I'm not really sure about this one, but I think that `loc` can also also be of type `Position`. At least in ESLint this is certainly the case.

**Comment by nzakas** [comment]
> I'm not really sure about this one, but I think that `loc` can also also be of type `Position`.

This is true based on [this code](https://github.com/eslint/eslint/blob/dab5478e8628447dbf9eaaa8b6f36d7ca253ed48/lib/linter/report-translator.js#L93-L98), but I'm pretty sure this was an error recovery mechanism rather than an expected practice. We want to encourage people to use full locations for reporting errors, otherwise you just get a single character underlined in an editor, which is difficult to see. Remember, these types were ported from `@types/eslint` as a starting point, and no one consulted us one what we wanted to encourage or discourage.

I'll make the adjustments in `@eslint/core`, but I think we should be looking to move our types to something that is stricter over time.
